### PR TITLE
Support request specific TLS configuration

### PR DESCRIPTION
**Pull request by madsodgaard**
Adds support for request-specific TLS configuration:
`Request(url: "https://webserver.com", tlsConfiguration: .forClient())`

https://github.com/apple/swift-nio-ssl/pull/280 must be released, before this can be merged.

### CODE REVIEW

**Comment by Lukasa** [comment]
@swift-server-bot add to allowlist

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift [hunk 1]
```diff
@@ -2916,4 +2916,22 @@ class HTTPClientTests: XCTestCase {
             XCTAssertEqual(error as? ExpectedError, .expected)
         }
     }
+    
+    func testRequestSpecificTLS() throws {
```
Comment (weissi) [inline Tests/AsyncHTTPClientTests/HTTPClientTests.swift:2920]: could we make a second request in this test and check the `RequestInfo` that the test server should return.

The `RequestInfo` has the following fields:

```swift
internal struct RequestInfo: Codable {
    var data: String
    var requestNumber: Int
    var connectionNumber: Int
}
```

so you should be able to write

```swift
        guard let response = maybeResponse, let body = response.body else {
            XCTFail("request failed")
            return
        }
        let connectionNumber = try? JSONDecoder().decode(RequestInfo.self, from: body).connectionNumber
```

and then (in a few test cases) we can validate that:

- two subsequent requests with _different_ `TLSConfiguration`s get different `connectionNumber`s
- two subsequent requests with the _same_ `TLSConfiguration` get the same `connectionNumber` for both requests.

File: Sources/AsyncHTTPClient/ConnectionPool.swift [hunk 2]
```diff
@@ -139,12 +139,16 @@ final class ConnectionPool {
             self.port = request.port
             self.host = request.host
             self.unixPath = request.socketPath
+            if let tls = request.tlsConfiguration {
```
Comment (weissi) [inline Sources/AsyncHTTPClient/ConnectionPool.swift:142]: @artemredkin should we add a test case that targets the pool directly here? Ie no actual connections?

File: Package.swift [hunk 3]
```diff
@@ -22,7 +22,8 @@ let package = Package(
     ],
     dependencies: [
         .package(url: "https://github.com/apple/swift-nio.git", from: "2.27.0"),
-        .package(url: "https://github.com/apple/swift-nio-ssl.git", from: "2.12.0"),
+        .package(url: "https://github.com/apple/swift-nio-ssl.git", .branch("main")),
+//        .package(url: "https://github.com/apple/swift-nio-ssl.git", from: "2.12.0"),
```
Comment (weissi) [inline Package.swift:26]: once NIOSSL is released, we need to change this back (and up the version number). Just as a reminder.

**Review by weissi — CHANGES_REQUESTED** [review]
Fab, thank you! This looks like a great start!

I left a few comments, mostly requesting changes that we don't accidentally merge the branch dependency.

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift (hunk 1 above)
Comment (madsodgaard) [reply Tests/AsyncHTTPClientTests/HTTPClientTests.swift:2920]: Sure! Did not spot that the we could test it that way, good to know! I added the assertions to the same test, lmk if you meant something else!

**Comment by weissi** [comment]
@madsodgaard hmm, weird 5.0 compiler warning (that we turn into errors):

```
[449/452] Compiling Swift Module 'AsyncHTTPClient' (15 sources)
/code/Sources/AsyncHTTPClient/BestEffortHashableTLSConfiguration.swift:16:18: error: result of call to 'bestEffortEquals' is unused
        lhs.base.bestEffortEquals(rhs.base)
                 ^               ~~~~~~~~~~
Build step 'Execute shell' marked build as failure
```

Also, async-http-client uses automatic format checking with SwiftFormat. If you run SwiftFormat over the source, then the "soundness" part should pass.

**Comment by madsodgaard** [comment]
@weissi Whoops, forgot a return statement 😅

**Comment by weissi** [comment]
@madsodgaard also the API breakage checker detected an API breakage:

```
08:37:54 /* Removed Decls */
08:37:54 Constructor HTTPClient.Request.init(url:method:headers:body:) has been removed
```

Note that in Swift you cannot just add a new parameter (even if it has a default value) without breaking API. Instead of adding the `tlsConfiguration:` parameter with a default value, you'll need to add a new `init` with an extra (non-defaulted) parameter. The existing `HTTPClient.Request.init(url:method:headers:body:)` should then just call the new `init` with the extra parameter.

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift (hunk 1 above)
Comment (weissi) [reply Tests/AsyncHTTPClientTests/HTTPClientTests.swift:2920]: thank you!

File: Sources/AsyncHTTPClient/Utils.swift [hunk 4]
```diff
@@ -150,17 +150,22 @@ extension NIOClientTCPBootstrap {
         let key = destination
 
         let requiresTLS = key.scheme.requiresTLS
+        // Override optional connection pool configuration.
+        var keyConfiguration = configuration
```
[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/AsyncHTTPClient/Utils.swift:154]: Also, should this merge perhaps be done at a higher level, say when we create the `HTTP1ConnectionProvider`? I am a bit nervous about having two separate configs from the perspective of the connection provider: it should always be creating the exact same connection each time.

File: Package.swift (hunk 3 above)
Comment (weissi) [reply Package.swift:26]: @madsodgaard we got them release https://github.com/apple/swift-nio-ssl/releases/tag/2.13.0

**Review by Lukasa — APPROVED** [review]
Cool, this LGTM.

**Review by weissi — APPROVED** [review]
Fantastic, thank you very much!

**Pull request merged**